File: frontend/src/lib/validation/clinical-ranges.ts
/**
 * Clinical reference ranges for vital signs validation.
 * Ranges are based on standard adult clinical guidelines.
 */

export type WarningLevel = "normal" | "warning" | "critical";

export interface VitalRange {
  min: number;
  max: number;
  warningLow: number;
  warningHigh: number;
  criticalLow: number;
  criticalHigh: number;
  unit: string;
  normalRange: string;
}

export const clinicalRanges = {
  systolic: {
    min: 40,
    max: 250,
    warningLow: 90,
    warningHigh: 140,
    criticalLow: 70,
    criticalHigh: 180,
    unit: "mmHg",
    normalRange: "90-140 mmHg",
  },
  diastolic: {
    min: 20,
    max: 150,
    warningLow: 60,
    warningHigh: 90,
    criticalLow: 40,
    criticalHigh: 120,
    unit: "mmHg",
    normalRange: "60-90 mmHg",
  },
  heartRate: {
    min: 30,
    max: 250,
    warningLow: 60,
    warningHigh: 100,
    criticalLow: 40,
    criticalHigh: 150,
    unit: "bpm",
    normalRange: "60-100 bpm",
  },
  temperature: {
    min: 90,
    max: 110,
    warningLow: 97,
    warningHigh: 99.5,
    criticalLow: 95,
    criticalHigh: 104,
    unit: "°F",
    normalRange: "97-99.5 °F",
  },
  respiratoryRate: {
    min: 4,
    max: 60,
    warningLow: 12,
    warningHigh: 20,
    criticalLow: 8,
    criticalHigh: 30,
    unit: "/min",
    normalRange: "12-20 /min",
  },
  oxygenSaturation: {
    min: 50,
    max: 100,
    warningLow: 95,
    warningHigh: 100,
    criticalLow: 90,
    criticalHigh: 100,
    unit: "%",
    normalRange: "95-100%",
  },
  weight: {
    min: 1,
    max: 1000,
    warningLow: 0,
    warningHigh: 1000,
    criticalLow: 0,
    criticalHigh: 1000,
    unit: "lbs",
    normalRange: "N/A",
  },
} as const satisfies Record<string, VitalRange>;

export type VitalType = keyof typeof clinicalRanges;

/**
 * Evaluates a vital sign value and returns the warning level.
 */
export function evaluateVital(
  vitalType: VitalType,
  value: number
): WarningLevel {
  const range = clinicalRanges[vitalType];

  // For oxygen saturation, we only care about low values
  if (vitalType === "oxygenSaturation") {
    if (value < range.criticalLow) return "critical";
    if (value < range.warningLow) return "warning";
    return "normal";
  }

  // For weight, no warning/critical ranges
  if (vitalType === "weight") {
    return "normal";
  }

  // For other vitals, check both low and high
  if (value < range.criticalLow || value > range.criticalHigh)
    return "critical";
  if (value < range.warningLow || value > range.warningHigh) return "warning";
  return "normal";
}

/**
 * Gets the appropriate CSS classes for a warning level.
 */
export function getWarningClasses(level: WarningLevel): string {
  switch (level) {
    case "critical":
      return "border-error text-error bg-error/10";
    case "warning":
      return "border-warning text-warning bg-warning/10";
    default:
      return "";
  }
}

/**
 * Gets the badge classes for a warning level.
 */
export function getWarningBadgeClasses(level: WarningLevel): string {
  switch (level) {
    case "critical":
      return "badge-error";
    case "warning":
      return "badge-warning";
    default:
      return "badge-ghost";
  }
}

/**
 * Generates a warning message for a vital sign.
 */
export function getWarningMessage(
  vitalType: VitalType,
  value: number,
  level: WarningLevel
): string | null {
  if (level === "normal") return null;

  const range = clinicalRanges[vitalType];
  const vitalNames: Record<VitalType, string> = {
    systolic: "Systolic BP",
    diastolic: "Diastolic BP",
    heartRate: "Heart rate",
    temperature: "Temperature",
    respiratoryRate: "Respiratory rate",
    oxygenSaturation: "Oxygen saturation",
    weight: "Weight",
  };

  const name = vitalNames[vitalType];
  const severity =
    level === "critical" ? "critically abnormal" : "outside normal range";

  return `${name} ${value} ${range.unit} is ${severity}. Normal: ${range.normalRange}`;
}

File: FhirHubServer/src/FhirHubServer.Api/Features/PatientManagement/DTOs/ClinicalRanges.cs
namespace FhirHubServer.Api.Features.PatientManagement.DTOs;

/// <summary>
/// Clinical reference ranges for vital signs validation.
/// Ranges are based on standard adult clinical guidelines.
/// </summary>
public static class ClinicalRanges
{
    public static class Systolic
    {
        public const decimal Min = 40;
        public const decimal Max = 250;
        public const decimal WarningLow = 90;
        public const decimal WarningHigh = 140;
        public const decimal CriticalLow = 70;
        public const decimal CriticalHigh = 180;
        public const string Unit = "mmHg";
        public const string NormalRange = "90-140 mmHg";
    }

    public static class Diastolic
    {
        public const decimal Min = 20;
        public const decimal Max = 150;
        public const decimal WarningLow = 60;
        public const decimal WarningHigh = 90;
        public const decimal CriticalLow = 40;
        public const decimal CriticalHigh = 120;
        public const string Unit = "mmHg";
        public const string NormalRange = "60-90 mmHg";
    }

    public static class HeartRate
    {
        public const decimal Min = 30;
        public const decimal Max = 250;
        public const decimal WarningLow = 60;
        public const decimal WarningHigh = 100;
        public const decimal CriticalLow = 40;
        public const decimal CriticalHigh = 150;
        public const string Unit = "bpm";
        public const string NormalRange = "60-100 bpm";
    }

    public static class Temperature
    {
        public const decimal Min = 90;
        public const decimal Max = 110;
        public const decimal WarningLow = 97;
        public const decimal WarningHigh = 99.5m;
        public const decimal CriticalLow = 95;
        public const decimal CriticalHigh = 104;
        public const string Unit = "°F";
        public const string NormalRange = "97-99.5 °F";
    }

    public static class RespiratoryRate
    {
        public const decimal Min = 4;
        public const decimal Max = 60;
        public const decimal WarningLow = 12;
        public const decimal WarningHigh = 20;
        public const decimal CriticalLow = 8;
        public const decimal CriticalHigh = 30;
        public const string Unit = "/min";
        public const string NormalRange = "12-20 /min";
    }

    public static class OxygenSaturation
    {
        public const decimal Min = 50;
        public const decimal Max = 100;
        public const decimal WarningLow = 95;
        public const decimal WarningHigh = 100;
        public const decimal CriticalLow = 90;
        public const decimal CriticalHigh = 100;
        public const string Unit = "%";
        public const string NormalRange = "95-100%";
    }

    public static class Weight
    {
        public const decimal Min = 1;
        public const decimal Max = 1000;
        public const string Unit = "lbs";
    }

    /// <summary>
    /// Evaluates a vital sign value and returns the warning level.
    /// </summary>
    public static WarningLevel EvaluateVital(string vitalType, decimal value)
    {
        return vitalType.ToLowerInvariant() switch
        {
            "systolic" => EvaluateRange(value, Systolic.CriticalLow, Systolic.CriticalHigh, Systolic.WarningLow, Systolic.WarningHigh),
            "diastolic" => EvaluateRange(value, Diastolic.CriticalLow, Diastolic.CriticalHigh, Diastolic.WarningLow, Diastolic.WarningHigh),
            "heartrate" => EvaluateRange(value, HeartRate.CriticalLow, HeartRate.CriticalHigh, HeartRate.WarningLow, HeartRate.WarningHigh),
            "temperature" => EvaluateRange(value, Temperature.CriticalLow, Temperature.CriticalHigh, Temperature.WarningLow, Temperature.WarningHigh),
            "respiratoryrate" => EvaluateRange(value, RespiratoryRate.CriticalLow, RespiratoryRate.CriticalHigh, RespiratoryRate.WarningLow, RespiratoryRate.WarningHigh),
            "oxygensaturation" => EvaluateLowOnly(value, OxygenSaturation.CriticalLow, OxygenSaturation.WarningLow),
            _ => WarningLevel.Normal
        };
    }

    private static WarningLevel EvaluateRange(decimal value, decimal criticalLow, decimal criticalHigh, decimal warningLow, decimal warningHigh)
    {
        if (value < criticalLow || value > criticalHigh)
            return WarningLevel.Critical;
        if (value < warningLow || value > warningHigh)
            return WarningLevel.Warning;
        return WarningLevel.Normal;
    }

    private static WarningLevel EvaluateLowOnly(decimal value, decimal criticalLow, decimal warningLow)
    {
        if (value < criticalLow)
            return WarningLevel.Critical;
        if (value < warningLow)
            return WarningLevel.Warning;
        return WarningLevel.Normal;
    }
}

public enum WarningLevel
{
    Normal,
    Warning,
    Critical
}
